namui-cli/src/build/build_server/code_watcher.rs

```diff
@@ -0,0 +1,164 @@
+            panic!("watch failed {:?}", error);
+        }
+
+        let context_clone = context.clone();
+        thread::spawn(move || -> ! {
+            loop {
+                match watcher_receiver.recv() {
+                    Ok(event) => {
+                        match event {
+                            DebouncedEvent::Create(_)
+                            | DebouncedEvent::Remove(_)
+                            | DebouncedEvent::Rename(_, _)
+                            | DebouncedEvent::Write(_) => {
+                                debug_println!("CodeWatcher: locking code_watcher.context...");
+                                let mut context = block_on(context_clone.lock());
+                                debug_println!("CodeWatcher: code_watcher.context locked");
+
+                                match context.sender.take() {
+                                    Some(sender) => {
+                                        if let Err(_) = sender.send(()) {
+                                            panic!("send failed in watcher")
+                                        }
+                                    }
+                                    None => {
+                                        context.was_changed = true;
+                                    }
+                                }
+                            }
+                            _ => (),
+                        };
+                    }
+                    Err(error) => eprintln!("{:?}", error),
+                }
+            }
+        });
+
+        Self {
+            watcher,
+            context,
+            watching_paths: HashSet::new(),
+            manifest_path,
+        }
+    }
+
+    pub async fn wait_for_change(&self) {
+        match self.check_change_state().await {
+            CodeWatcherChangeState::Changed => (),
+            CodeWatcherChangeState::NotChanged(receiver) => match receiver.await {
+                Ok(_) => (),
+                Err(error) => panic!("code_watcher oneshot receive error: {:?}", error),
+            },
+        }
+    }
+
+    async fn check_change_state(&self) -> CodeWatcherChangeState {
+        debug_println!("check_changed: locking code_watcher.context...");
+        let mut context = self.context.lock().await;
+        debug_println!("check_changed: code_watcher.context locked");
+        match context.was_changed {
+            true => {
+                context.was_changed = false;
+                CodeWatcherChangeState::Changed
+            }
+            false => {
+                let (sender, receiver) = oneshot::channel::<()>();
+                match context.sender {
+                    Some(_) => unreachable!(),
+                    None => context.sender = Some(sender),
+                }
+                CodeWatcherChangeState::NotChanged(receiver)
+            }
+        }
+    }
+
+    pub fn update_watching_paths(&mut self) {
```

Comment:
I think it is kind of initialization, so it would be better to make it private function and call in `new`.

Reply:
Ah I got it. It should be called if Cargo.toml is changed. ok.